pyro/infer/rws.py

```diff
@@ -0,0 +1,227 @@
+
+        if self.insomnia < 1:
+            # compute sleep phi loss
+            _model = pyro.poutine.uncondition(model)
+            _guide = guide
+            _log_q = 0.
+
+            if self.vectorize_particles:
+                if self.max_plate_nesting == float('inf'):
+                    self._guess_max_plate_nesting(_model, _guide, *args, **kwargs)
+                _model = self._vectorized_num_sleep_particles(_model)
+                _guide = self._vectorized_num_sleep_particles(guide)
+
+            for _ in range(1 if self.vectorize_particles else self.num_sleep_particles):
+                _model_trace = poutine.trace(_model).get_trace(*args, **kwargs).detach()
+                _guide_trace = self._get_matched_trace(_model_trace, _guide, *args, **kwargs)
+                _log_q += _guide_trace.log_prob_sum()
+
+            sleep_phi_loss = -_log_q / self.num_sleep_particles
+            warn_if_nan(sleep_phi_loss, "sleep phi loss")
+
+        # compute phi loss
+        phi_loss = sleep_phi_loss if self.insomnia == 0 \
+            else wake_phi_loss if self.insomnia == 1 \
+            else self.insomnia * wake_phi_loss + (1. - self.insomnia) * sleep_phi_loss
+
+        return wake_theta_loss, phi_loss
+
+    def loss(self, model, guide, *args, **kwargs):
+        """
+        :returns: returns model loss and guide loss
+        :rtype: float, float
+
+        Computes the re-weighted wake-sleep estimators for the model (wake-theta) and the
+          guide (insomnia * wake-phi + (1 - insomnia) * sleep-phi).
+        """
+        with torch.no_grad():
+            wake_theta_loss, phi_loss = self._loss(model, guide, *args, **kwargs)
+
+        return wake_theta_loss, phi_loss
+
+    def loss_and_grads(self, model, guide, *args, **kwargs):
+        """
+        :returns: returns model loss and guide loss
+        :rtype: float
+
+        Computes the RWS estimators for the model (wake-theta) and the guide (wake-phi).
+        Performs backward as appropriate on both, using num_particle many samples/particles.
+        """
+        wake_theta_loss, phi_loss = self._loss(model, guide, *args, **kwargs)
+        try:
+            wake_theta_loss.backward(retain_graph=True)
+        except RuntimeError:
+            pass
+        phi_loss.backward()
+
+        return wake_theta_loss.detach(), phi_loss.detach()
+
+    def _vectorized_num_sleep_particles(self, fn):
+        """
+        Copy of `_vectorised_num_particles` that uses `num_sleep_particles`.
+        """
+        def wrapped_fn(*args, **kwargs):
+            if self.num_sleep_particles == 1:
+                return fn(*args, **kwargs)
+            with pyro.plate("num_sleep_particles_vectorized", self.num_sleep_particles, dim=-self.max_plate_nesting):
+                return fn(*args, **kwargs)
+
+        return wrapped_fn
+
+    @staticmethod
+    def _get_matched_trace(model_trace, guide, *args, **kwargs):
+        # TODO: hardcoded kwarg 'observations'?
```

Comment:
So, I'm not quite sure how we could avoid it, really.

Basically, when doing _sleep-phi_,  we need to switch the processing order from _wake-*_; we first sample the (unconditioned) model trace, and then replay that on the guide. 

Because the guide doesn't `sample` the observation variable, but is instead given it, we need to match things up somehow for the `replay`. We hence using the same procedure as with `CSIS` by forcing use of a named keyword `observations` to hold the observations that the guide uses.

Reply:
Oh, I don't understand the CSIS use of `kwargs["observations"]` either. @eb8680 can you explain?

@iffsid can you explain the TODO?

Reply:
The guide distribution `q(z | x)` depends in general on the values of observed variables `x` in the model's joint distribution `p(x, z)`.  In the wake phase, those values are given as data, but in the sleep phase, the data is ignored and those values are instead drawn from the model `p(x | z)`, and need to be extracted from a model trace and passed to the guide.  They can't currently be passed with an effect handler because the dependence of `q` on `x` is deterministic and arbitrary.  I suppose after #2196 we could update `CSIS` and `RWS` to require the guide to have `deterministic` calls around the values `x` and accomplish the same thing with `condition`, although that doesn't seem any more elegant than the current solution.

Reply:
@fritzo The TODO was just to ensure that we explicitly discussed this issue in the PR. :)
I'll drop the TODO comment and add a note in the docs about the explicit `observations` keyword as well.

Reply:
Done in `7532af20`.